fili-core/src/main/java/com/yahoo/bard/webservice/table/LogicalTableSchema.java

```diff
@@ -45,10 +49,30 @@ public LogicalTableSchema(TableGroup tableGroup, Granularity granularity, Metric
         return Stream.concat(
                 tableGroup.getDimensions().stream()
                 .map(DimensionColumn::new),
-                tableGroup.getApiMetricNames().stream()
-                        .filter(apiMetricName ->  apiMetricName.isValidFor(granularity))
-                        .map(ApiMetricName::getApiName)
-                        .map(name -> new LogicalMetricColumn(name, metricDictionary.get(name)))
+                buildMetricColumns(tableGroup.getApiMetricNames(), granularity, metricDictionary)
         ).collect(Collectors.toCollection(LinkedHashSet::new));
     }
+
+    /**
+     * Buid a stream of metric columns by filtering the table group's collection of ApiMetricNames.
+     * Metric names are resolved and then filtered considering the table granularity and the metric referenced.
+     *
+     * @param apiMetricNames  The names of the apiMetrics being bound and filtered
+     * @param granularity  The grain used to filter those metric names
+     * @param metricDictionary  The dictionary to resolve the logical metric instances from
+     *
+     * @return A stream of metric columns, filtered for compatibility with the grain.
+     */
+    private static Stream<MetricColumn> buildMetricColumns(
+            Collection<ApiMetricName> apiMetricNames,
+            Granularity granularity,
+            MetricDictionary metricDictionary
+    ) {
+        return apiMetricNames.stream()
+                .collect(Collectors.toMap(Function.identity(), entry -> metricDictionary.get(entry.getApiName())))
+                .entrySet().stream()
+                .filter(entry -> entry.getKey().isValidFor(entry.getValue(), granularity))
+                .map(Map.Entry::getValue)
```

Comment:
Can this be simplified? I don't _think_ we need to build a `Map<ApiMetricName, LogicalMetric>`, then filter by the keys, only to then grab the values. I _think_ we could just filter out the APIMetricNames, and then convert the remaining ones to LogicalMetrics. I've not tested this code change though.